### PR TITLE
Inlining lambdas and removing unused definitions

### LINKED ISSUE #283: [BUG] Inliner fails to inline operators
Somewhat unexpected, the following simple piece of code does not work:

```tla
----------------------------- MODULE Callback ----------------------------      
EXTENDS Integers

VARIABLES x

Pick(Cb(_)) ==
    \E i \in 1..10:
        Cb(i)

Init ==
    LET Cb(i) ==
        x = i
    IN
    Pick(Cb)

Next ==
    LET Cb(i) ==
        x' = i
    IN
    Pick(Cb)

Inv ==
    x > 0
==========================================================================
```

### PR DESCRIPTION
**Pull request by konnov**
This PR fixes two issues that come together, as it is hard to test one without the other:

 - [x] Fix #283: unused operator definitions are removed after inlining. Otherwise, `FlatLanguagePred` produces an error on high-level operators.
 - [x] Replace applications of lambda expressions like `(LAMBDA x: e)(y)` with `e[y/x]`. Such expressions do not occur in TLA+ itself, but are produced by our inlining. For instance, lambda operators appeared in #285.

### CODE REVIEW

File: tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/passes/ConfigurationPassImpl.scala [hunk 1]
```diff
@@ -47,15 +47,24 @@ class ConfigurationPassImpl @Inject()(val options: WriteablePassOptions,
...
+    setFallbackOptions(relevantOptions)
 
     // make sure that the required operators are defined
-    ensureDeclarationsArePresent(currentModule)
+    ensureDeclarationsArePresent(currentModule, relevantOptions)
+
+    // copy the relevant options back in options
+    for (name <- NormalizedNames.STANDARD_OPTION_NAMES) {
+      relevantOptions.get[Any]("checker", name)
+        .collect { case value => options.set("checker." + name, value) }
```
Comment (shonfeder) [inline tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/passes/ConfigurationPassImpl.scala:61]: This is identical to the body of `copyRelevantOptions`, pace a parameter or two. Would it make more sense to abstract into a single function that is called two different times with different arguments? I may be off the mark, but I'm finding the logic hard to follow here. Seems like a lot of stateful mutations going on for something that appears to be a mere matter of filtering out members of a collection and verifying that they have certain properties.

**Review by shonfeder — APPROVED** [review]
I'm afraid I don't have a super clear grasp of the motivating bugs or how this address it, but I've tried to offer some feedback that might be useful in any case.

File: mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/WriteablePassOptions.scala [hunk 2]
```diff
@@ -42,7 +42,12 @@ class WriteablePassOptions extends PassOptions {
     */
   def get[T](passName: String, optionName: String): Option[T] = {
     store.get(passName + "." + optionName) match {
-      case Some(value) => Some(value.asInstanceOf[T])
+      case Some(value: T) =>
+        Some(value)
+
+      case Some(value) =>
+        throw new IllegalArgumentException(s"Option $optionName has unexpected type: " + value.getClass)
```
[the remark this answers is not in the document]
Comment (konnov) [reply mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/WriteablePassOptions.scala:49]: It is more of a convenience issue. In principle, the map stores objects of any kind. Since we do not want to do `isInstanceOf` all the time, this check is isolated in the method `Options.get[T]`. I guess, one can offload the whole work to the compiler. However, I am not sure it is worth the effort in this case.

File: test/tla/cli-integration-tests.md [hunk 3]
```diff
@@ -407,6 +407,19 @@ EXITCODE: ERROR (99)
 [99]
 ```
 
+### check Callback.tla succeeds
+
+This tests demonstrates that one can implement non-determinism with
+exists and use a callback to do an assignment to a variable.
+As it requires tricky operator inlining, here is the test.
```
[the remark this answers is not in the document]
Comment (konnov) [reply test/tla/cli-integration-tests.md:414]: It is also an interesting example that the users may want to see. The unit tests are more cryptic, as they are written in the intermediate language. Perhaps, we will that one day by introducing the text version of the IR.

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/config/CheckerExceptionAdapter.scala [hunk 4]
```diff
@@ -42,6 +42,9 @@ class CheckerExceptionAdapter @Inject()(sourceStore: SourceStore,
       val msg = "%s\n%s".format(err.getMessage, err.errors.map(ofTypeInferenceError).mkString("\n"))
       NormalErrorMessage(msg)
 
+    case err: NotInKeraError =>
```
[the remark this answers is not in the document]
Comment (konnov) [reply tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/config/CheckerExceptionAdapter.scala:45]: Right. I have updated the message

File: tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/passes/ConfigurationPassImpl.scala (hunk 1 above)
Comment (konnov) [reply tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/passes/ConfigurationPassImpl.scala:61]: That is a good catch. Fixed.

**Comment by konnov** [comment]
inline lambdas and fix #283 (filtered unused operators)

File: tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/passes/ConfigurationPassImpl.scala [hunk 5]
```diff
@@ -87,13 +85,12 @@ class ConfigurationPassImpl @Inject()(val options: WriteablePassOptions,
   }
 
   // copy the relevant options
-  private def copyRelevantOptions(): WriteablePassOptions = {
+  private def copyRelevantOptions(from: PassOptions, to: WriteablePassOptions): Unit = {
```
Comment (shonfeder) [inline tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/passes/ConfigurationPassImpl.scala:88]: Nice, yeah this definitely helps make the flow of information a bit clearer. Thanks :)
